gateway/src/node/storage/mod.rs

```diff
@@ -348,29 +390,112 @@ impl Storage for PersistentStorage {
         Ok(())
     }
 
+    async fn insert_credential(
+        &self,
+        credential: CredentialSpendingRequest,
+    ) -> Result<(), StorageError> {
+        self.credential_manager
+            .insert_credential(credential.to_bs58())
+            .await?;
+        Ok(())
+    }
+
+    async fn insert_pending_credential(
+        &self,
+        pending: PendingCredential,
+    ) -> Result<(), StorageError> {
+        self.credential_manager
+            .insert_pending_credential(
+                pending.credential.to_bs58(),
+                pending.address.into(),
+                pending
+                    .api_clients
+                    .iter()
+                    .map(|client| client.api_url().to_string())
+                    .collect::<Vec<_>>()
+                    .join(","),
+                pending
+                    .proposal_id
+                    .map(|id| id.to_string())
+                    .unwrap_or_default(),
+            )
+            .await?;
+        Ok(())
+    }
     async fn insert_spent_credential(
         &self,
-        blinded_serial_number: BlindedSerialNumber,
+        blinded_serial_number: String,
         was_freepass: bool,
         client_address: DestinationAddressBytes,
     ) -> Result<(), StorageError> {
         self.bandwidth_manager
             .insert_spent_credential(
-                &blinded_serial_number.to_bs58(),
+                &blinded_serial_number,
                 was_freepass,
                 &client_address.as_base58_string(),
             )
             .await?;
         Ok(())
     }
 
+    async fn remove_pending_credential(&self, id: i64) -> Result<(), StorageError> {
+        self.credential_manager
+            .remove_pending_credential(id)
+            .await?;
+        Ok(())
+    }
+
+    async fn get_all_pending_credential(
+        &self,
+    ) -> Result<Vec<(i64, PendingCredential)>, StorageError> {
```

Comment:
I feel I've already left plenty of comments about this one : )

Reply:
It's literally just an id